pkg/manager/manager.go

```diff
@@ -341,6 +342,34 @@ type LeaderElectionRunnable interface {
 	NeedLeaderElection() bool
 }
 
+// NewCancelRunnable returns a new Runnable and a function to cancel that runnable.
+func NewCancelRunnable(r Runnable) (Runnable, context.CancelFunc) {
+	mu := sync.Mutex{}
+	var cancels []func()
+	canceled := false
+
+	return RunnableFunc(func(ctx context.Context) error {
+			mu.Lock()
+			var cancel context.CancelFunc
+			ctx, cancel = context.WithCancel(ctx)
+			if canceled {
+				cancel()
+			} else {
+				cancels = append(cancels, cancel)
+			}
+			mu.Unlock()
+
+			return r.Start(ctx)
+		}), func() {
+			mu.Lock()
+			defer mu.Unlock()
+			canceled = true
+			for _, cancel := range cancels {
+				cancel()
+			}
+		}
+}
+
```

Comment:
@inteon  As a user, how can i use this func to stop controller ?

Reply:
I added a test here: https://github.com/kubernetes-sigs/controller-runtime/pull/2159/files#diff-21b67ad2c12fb31030a5f2e93e1f1e1593b7375402163a660bc056e126cb0cfbR349-R372